Parse DNSSEC AD and CD bits

micrictor opened the pull request:
Parse authentic data (AD) and checking disabled (CD) bits according to RFC 2535. Leaves the Z field as-is, in case users are already handling this elsewhere and depend on the value being the integer for all 3 bits.

https://www.rfc-editor.org/rfc/rfc2535#section-6.1

Fixes #2672

timwoj commented on scripts/base/protocols/dns/main.zeek:
```diff
@@ -429,6 +443,14 @@ event dns_request(c: connection, msg: dns_msg, query: string, qtype: count, qcla
 	c$dns$qtype_name  = query_types[qtype];
 	c$dns$Z           = msg$Z;
 
```
Does AD get parsed in a request? I only did a quick skim of the RF, but it sounds like it should only be valid in a response.

micrictor replied:
In the canonical RFC for AD, it was left undefined; RFC 6840 clarified that the AD bit can be set by the request or response
https://www.rfc-editor.org/rfc/rfc6840#section-5.7

timwoj commented:
Looks good to me. Can you update `testing/external/commit-hash.zeek-testing` to the commit from your PR on that repo, and re-push this for the CI coverage? I have to imagine that it'll still fail because zeek-testing-private fails but I'll have to update that part.

awelzel commented on scripts/base/protocols/dns/main.zeek:
```diff
@@ -359,6 +367,13 @@ hook DNS::do_reply(c: connection, msg: dns_msg, ans: dns_answer, reply: string)
 
+	# Decode AD bit according to RFC 2535 if Z is set.
+	# Leave Z untouched in case existing users post-process the number value.
+	if ( msg$Z != 0 ) 
```
IMO we should adapt the analyzer to decode the bits and then extend `dns_msg` with AD and CD fields (possibly always set to whatever Z would evaluate to rather than making them optional). This would give script writers immediate access to the fields without relying on logic in `dns/main.zeek`.

Extending dns.log with these new fields could be discussed separately (not sure the original ticket assumed that would be done). These fields might be useful enough in every deployment, but could also see not adding them (or not marking them `&log`) by default. But suggest to add them for now.

micrictor replied:
Decoding the bits directly in the parser SGTM, and makes sense as these flags should be parsed out the same as RA/RD/etc - will make that change. Unless someone thinks otherwise, I'll still leave `Z` untouched, on the off chance someone downwstream depends on it holding all three bits. 

I'll still add them to the log - IMO, we log every other standard part of the DNS header, so we should log these as well.

ckreibich replied:
I agree with all here but this last bit is unusual — we don't normally leave a field in a message header and then parse out its bits in addition. I would normally vote for deprecation of the old `Z` field and introduction of a new one. On the other hand this case seems a bit special because even popular DNS resources often don't mention AD & CD.

 So ... in the end I agree — let's keep the Z field as-is, but let's also be very clear in the docstrings about how Z, AD, and CD relate in our implementation.

micrictor replied:
Done - newly pushed changes moved the bit-parsing to the protocol parser. Also pushed a change to docs that explicitly calls out that Z is the 3-bit Z specified in the original DNS RFC.

awelzel commented:
@micrictor - are you still interested and planning in addressing the feedback?

micrictor commented:
Yes! Sorry about the delay - life is currently happening to me at an above
average rate, it seems.



On Mon, Feb 27, 2023 at 12:55 AM Arne Welzel ***@***.***>
wrote:

> @micrictor <https://github.com/micrictor> - are you still interested and
> planning in addressing the feedback?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zeek/zeek/pull/2772#issuecomment-1445934696>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTCAGMQXNXBXL4ET2WZV6DWZRTYHANCNFSM6AAAAAAUY4DIXI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

awelzel commented on scripts/base/protocols/dns/main.zeek:
```diff
@@ -428,6 +443,8 @@ event dns_request(c: connection, msg: dns_msg, query: string, qtype: count, qcla
 	c$dns$qtype       = qtype;
 	c$dns$qtype_name  = query_types[qtype];
 	c$dns$Z           = msg$Z;
+	c$dns$AD          = msg$AD;
```
I'm unsure if we should set `AD` when processing `dns_request`: In requests it should be 0 IIUC.

We could instead consider to flag AD being set in requests (or CD being set in responses) as a weird, but can do as a follow up.

micrictor replied:
RFC 6840 states that the AD bit can be set by the request or response
https://www.rfc-editor.org/rfc/rfc6840#section-5.7

micrictor replied:
More info on why - The value set here will be overriden by the response value if both the request and reply are seen by Zeek. Definitely worth adding weird logs in the future.

awelzel replied:
Thanks for re-linking that RFC.

> request and reply are seen by Zeek

Which should be the common case. What's there now I think is okay. As you mention, the `AD` column in the log will be overwritten by the response's AD flag and that's semi great.

I'm paging @ckreibich if we should solve this somehow differently in a follow-up (new column, or let script writers extend the log themselves), but I'll look into merging your this changeset as is.

awelzel commented on scripts/base/protocols/dns/main.zeek:
```diff
@@ -60,9 +60,17 @@ export {
 		RA:            bool               &log &default=F;
-		## A reserved field that is usually zero in
-		## queries and responses.
+		## A reserved field that is usually zero in queries and
```
Should we drop the comment as it is not applicable today anymore?

micrictor replied:
Updated comment to match the zeek-docs, let me know if that addresses your concern

awelzel commented on testing/btest/btest-results.xml:
```diff
@@ -0,0 +1,1503 @@
+<?xml version="1.0" ?>
```
Mind dropping that file and the tmp.tar.gz file below again?

awelzel reviewed (commented):
Hey @micrictor - I had a few more comments and it seems you added some unrelated files (and a merge commit is in the history, too). There's something off with the external testing baselines, too.

If you agree with the comments and would rather hand this off, I'm happy to pick it up and see it through. I'd squash the history a bit, but keep your authorship, of course.

Let us know.
